scene/2d/animated_sprite_2d.cpp

```diff
@@ -175,21 +175,21 @@ void AnimatedSprite2D::_notification(int p_what) {
 				if (timeout <= 0) {
 					timeout = _get_frame_duration();
 
-					int fc = frames->get_frame_count(animation);
-					if ((!backwards && frame >= fc - 1) || (backwards && frame <= 0)) {
+					int last_frame = frames->get_frame_count(animation) - 1;
+					if ((!backwards && frame >= last_frame) || (backwards && frame <= 0)) {
 						if (frames->get_animation_loop(animation)) {
-							if (backwards) {
-								frame = fc - 1;
-							} else {
+							if (!backwards) {
 								frame = 0;
+							} else {
+								frame = last_frame;
 							}
 
 							emit_signal(SceneStringNames::get_singleton()->animation_finished);
 						} else {
-							if (backwards) {
-								frame = 0;
+							if (!backwards) {
+								frame = last_frame;
 							} else {
-								frame = fc - 1;
+								frame = 0;
 							}
 
 							if (!is_over) {
```

Comment:
I recommend it should not complicate the code and increase the cost of processing a branch to shorten a few lines.

```cpp
if ((!backwards && frame >= last_frame) || (backwards && frame <= 0)) {
	if (frames->get_animation_loop(animation)) {
		if (!backwards) {
			frame = 0;
		} else {
			frame = last_frame;
		}

		emit_signal(SceneStringNames::get_singleton()->animation_finished);
	} else {
		if (!backwards) {
			frame = last_frame;
		} else {
			frame = 0;
		}

		if (!is_over) {
			is_over = true;
			emit_signal(SceneStringNames::get_singleton()->animation_finished);
		}
	}
} else {
	if (!backwards) {
		frame++;
	} else {
		frame--;
	}
}
```
↓
```cpp
if (!backwards) {
	if (frame >= last_frame) {
		if (frames->get_animation_loop(animation)) {
			frame = 0;
			emit_signal(SceneStringNames::get_singleton()->animation_finished);
		} else {
			frame = last_frame;
			if (!is_over) {
				is_over = true;
				emit_signal(SceneStringNames::get_singleton()->animation_finished);
			}
		}
	} else {
		frame++;
	}
} else {
	if (frame <= 0) {
		if (frames->get_animation_loop(animation)) {
			frame = last_frame;
			emit_signal(SceneStringNames::get_singleton()->animation_finished);
		} else {
			frame = 0;
			if (!is_over) {
				is_over = true;
				emit_signal(SceneStringNames::get_singleton()->animation_finished);
			}
		}
	} else {
		frame--;
	}
}
```

Reply:
I see it now.